### PR TITLE
Adds Keycard Authentication tgui

### PR DESCRIPTION
**Pull request by tkdrg**
Also rewrites keycard authentication while I'm at it. They were awful.

### CODE REVIEW

File: tgui/src/interfaces/keycard_auth.ract [hunk 1]
```diff
@@ -0,0 +1,20 @@
+{{#if data.broken}}
+	<ui-notice>
+		<span>This device looks broken. The display is dark and none of the buttons work.</span>
```
Comment (neersighted) [inline tgui/src/interfaces/keycard_auth.ract]: This is flavor text, not UI text. I wouldn't show the UI if broken.

File: tgui/src/interfaces/keycard_auth.ract [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+{{elseif data.waiting}}
+	<ui-notice>
+		<span>Waiting for another device to confirm your request...</span>
+	</ui-notice>
+{{else}}
+	<ui-display title="Keycard Authentication">
+		<ui-section>
+			{{#if data.authRequired}}
+				<ui-button icon='check' action="authSwipe">Authorize</ui-button>
+			{{else}}
```
Comment (neersighted) [inline tgui/src/interfaces/keycard_auth.ract]: we prefer snake_case to camelCase in tgui

File: code/modules/security levels/keycard authentication.dm [hunk 3]
```diff
@@ -53,93 +44,59 @@
...
-	sleep(confirm_delay)
-	if(confirmed)
-		confirmed = 0
-		trigger_event(event)
-		log_game("[key_name(event_triggered_by)] triggered and [key_name(event_confirmed_by)] confirmed event [event]")
-		message_admins("[key_name(event_triggered_by)] triggered and [key_name(event_confirmed_by)] confirmed event [event]")
-	reset()
-
-/obj/machinery/keycard_auth/proc/receive_request(obj/machinery/keycard_auth/source)
+/obj/machinery/keycard_auth/attack_hand(mob/user)
```
Comment (neersighted) [inline code/modules/security levels/keycard authentication.dm]: No need for this, the parent calls interact() for you

File: code/modules/security levels/keycard authentication.dm [hunk 4]
```diff
@@ -53,93 +44,59 @@
...
+	interact(user)
+
```
Comment (neersighted) [inline code/modules/security levels/keycard authentication.dm]: Write your own interact() proc that checks if the device is not broken

File: code/modules/security levels/keycard authentication.dm [hunk 5]
```diff
@@ -53,93 +44,59 @@
...
-		dat += "<li><A href='?src=\ref[src];triggerevent=Red alert'>Red alert</A></li>"
-		dat += "<li><A href='?src=\ref[src];triggerevent=Emergency Maintenance Access'>Emergency Maintenance Access</A></li>"
-		dat += "</ul>"
-		user << browse(dat, "window=keycard_auth;size=500x250")
-	if(screen == 2)
-		dat += "Please swipe your card to authorize the following event: <b>[event]</b>"
-		dat += "<p><A href='?src=\ref[src];reset=1'>Back</A>"
-		user << browse(dat, "window=keycard_auth;size=500x250")
-	return
+/obj/machinery/keycard_auth/get_ui_data()
```
Comment (neersighted) [inline code/modules/security levels/keycard authentication.dm]: Put this after ui_interact

File: code/modules/security levels/keycard authentication.dm [hunk 6]
```diff
@@ -53,93 +44,59 @@
...
+	var/list/data = list()
+	data["broken"] = (stat & (NOPOWER|BROKEN)) ? 1 : 0
```
Comment (neersighted) [inline code/modules/security levels/keycard authentication.dm]: This should be an interact() check

File: code/modules/security levels/keycard authentication.dm [hunk 7]
```diff
@@ -1,50 +1,41 @@
...
+	ev = keycard_events.addEvent("triggerEvent", src, "triggerEvent")
+
+/obj/machinery/keycard_auth/Destroy()
+	keycard_events.clearEvent("triggerEvent", ev)
+	qdel(ev)
+	. = ..()
 
 /obj/machinery/keycard_auth/attack_ai(mob/user)
 	if(IsAdminGhost(user))
 		return
```
Comment (neersighted) [inline code/modules/security levels/keycard authentication.dm]: This breaks now, should probably be ..()

**Comment by xxalpha** [comment]
pictures please

File: code/modules/security levels/keycard authentication.dm (hunk 4 above)
Comment (neersighted) [inline code/modules/security levels/keycard authentication.dm]: Ignore me, it should be in ui_status, eg

```
/obj/machinery/airalarm/ui_status(mob/user)
	if(isAI(user))
		user << "<span class='notice'>The station AI is not to interact with these devices.</span>"
	else if(!broken)
		return ..()
	return UI_CLOSE
```

File: code/modules/security levels/keycard authentication.dm (hunk 6 above)
Comment (neersighted) [inline code/modules/security levels/keycard authentication.dm]: No UI should pop up if broken, so you don't need this.

File: code/modules/security levels/keycard authentication.dm (hunk 3 above)
Comment (neersighted) [inline code/modules/security levels/keycard authentication.dm]: Remove any attack_ and interact() procs, the parent machinery procs will call ui_interact for you.

**Comment by tkdrg** [comment]
pics
2016-01-24-143937_376x144_scrot
2016-01-24-143950_1366x768_scrot

File: tgui/src/interfaces/keycard_auth.ract [hunk 8]
```diff
@@ -0,0 +1,16 @@
+{{#if data.waiting}}
+	<ui-notice>
+		<span>Waiting for another device to confirm your request...</span>
+	</ui-notice>
+{{else}}
+	<ui-display title="Keycard Authentication">
+		<ui-section>
+			{{#if data.auth_required}}
+				<ui-button icon='check' action="auth_swipe">Authorize {{data.auth_required}}</ui-button>
```
Comment (neersighted) [inline tgui/src/interfaces/keycard_auth.ract]: single quotes please

**Comment by tkdrg** [comment]
Should be good now hopefully

**Comment by PKPenguin321** [comment]
Ohhh, nice. :+1:

File: code/modules/security levels/keycard authentication.dm [hunk 9]
```diff
@@ -1,145 +1,99 @@
...
+	else if(isanimal(user))
+		user << "<span class='warning'>You are too primitive to use this device!</span>"
+	else
+		return ..()
+	return UI_CLOSE
 
-/obj/machinery/keycard_auth/Topic(href, href_list)
-	if(..())
+/obj/machinery/keycard_auth/ui_act(action, params)
+	if(waiting || !allowed(usr))
```
Comment (neersighted) [inline code/modules/security levels/keycard authentication.dm]: Needs a ..()

`if(..()|waiting|!allowed(usr))

File: code/modules/security levels/keycard authentication.dm [hunk 10]
```diff
@@ -1,145 +1,99 @@
...
-		dat += "</ul>"
-		user << browse(dat, "window=keycard_auth;size=500x250")
-	if(screen == 2)
-		dat += "Please swipe your card to authorize the following event: <b>[event]</b>"
-		dat += "<p><A href='?src=\ref[src];reset=1'>Back</A>"
-		user << browse(dat, "window=keycard_auth;size=500x250")
 	return
 
+/obj/machinery/keycard_auth/ui_interact(mob/user, ui_key = "main", datum/tgui/ui = null, force_open = 0, \
+					datum/tgui/master_ui = null, datum/ui_state/state = physical_state)
```
Comment (neersighted) [inline code/modules/security levels/keycard authentication.dm]: If state is physical_state, pretty sure AIs cannot interact anyway (just like canisters).

**Pull request merged**